### PR TITLE
add package @storybook/addon-knobs

### PR DESCRIPTION
**Pull request by joscha**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [ ] ~~The package does not provide its own types, and you can not add them.~~ It has been decided to move away from providing the types within the package as there is no owner using TS and the types were outdated most of the time, see https://github.com/storybooks/storybook/issues/1166 and https://github.com/storybooks/storybook/pull/1154
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] ~~Create it with `dts-gen --dt`, not by basing it on an existing project.~~ _(Isn't possible at the time, opened https://github.com/Microsoft/dts-gen/pull/63 which is merged, but has not been released, yet /cc @RyanCavanaugh)_
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

/cc @ndelangen
references #16970, https://github.com/storybooks/storybook/pull/1339

### CODE REVIEW

File: types/storybook__addon-knobs/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+// Project: https://github.com/storybooks/storybook
+// Definitions by: Joscha Feth <https://github.com/joscha>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.2
+
+/// <reference types="storybook__react" />
+
+// TODO: Once TS2.3 is released,
+// https://github.com/Microsoft/TypeScript/issues/14819 should be fixed.
+// Then, upgrade this package's typescript version to 2.3 and
```
Comment (sandersn) [inline types/storybook__addon-knobs/index.d.ts]: 2.3 is released so you can do this TODO.

Comment (sandersn) [inline types/storybook__addon-knobs/index.d.ts]: Might be nice to do it for `@storybook/react` too

Comment (joscha) [inline types/storybook__addon-knobs/index.d.ts]: @sandersn Is there an example of a scoped package that has done this already? Do you think we can do it as a follow-up and do both packages at once?

Comment (joscha) [inline types/storybook__addon-knobs/index.d.ts]: I just tried removing the wrappers, but now get:
```
/Users/joscha/Development/DefinitelyTyped/types/storybook__addon-knobs/index.d.ts
ERROR: 10:32  expect  TypeScript compile error: Cannot find module '@storybook/react'.

/Users/joscha/Development/DefinitelyTyped/types/storybook__addon-knobs/storybook__addon-knobs-tests.tsx
ERROR: 2:27   expect  TypeScript compile error: Cannot find module '@storybook/react'.
ERROR: 14:8   expect  TypeScript compile error: Cannot find module '@storybook/addon-knobs'.
```

I have used `/// <reference types="storybook__react" />` at the top of the file, but for some reason it still is not picked up correctly. Any idea what I might be doing wrong?

Comment (joscha) [inline types/storybook__addon-knobs/index.d.ts]: Tried a `rm -rf node_modules` and clean install to make sure its not an outdated `dtslint` version.

Full command:

```
$ npm run lint storybook__addon-knobs

> definitely-typed@0.0.1 lint /Users/joscha/Development/DefinitelyTyped
> dtslint types "storybook__addon-knobs"

Installing to /Users/joscha/Development/DefinitelyTyped/node_modules/dtslint/typescript-installs/2.3...
npm notice created a lockfile as package-lock.json. You should commit this file.

Installed!
/Users/joscha/Development/DefinitelyTyped/types/storybook__addon-knobs/index.d.ts
ERROR: 8:32  expect  TypeScript compile error: Cannot find module '@storybook/react'.

/Users/joscha/Development/DefinitelyTyped/types/storybook__addon-knobs/storybook__addon-knobs-tests.tsx
ERROR: 4:27  expect  TypeScript compile error: Cannot find module '@storybook/react'.
ERROR: 16:8  expect  TypeScript compile error: Cannot find module '@storybook/addon-knobs'.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! definitely-typed@0.0.1 lint: `dtslint types "storybook__addon-knobs"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the definitely-typed@0.0.1 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joscha/.npm/_logs/2017-06-22T21_00_18_487Z-debug.log
```

Comment (joscha) [inline types/storybook__addon-knobs/index.d.ts]: The branch with the `declare` removals diffs as following to this one here: https://github.com/joscha/DefinitelyTyped/compare/joscha/add-@storybook/addon-knobs...joscha:joscha/add-@storybook/addon-knobs-remove-declare?expand=1

Comment (sandersn) [inline types/storybook__addon-knobs/index.d.ts]: I think @andy-ms is the better one to ask since I think he made the name-mangling-lookup change.

**Comment by sandersn** [comment]
Let's merge this as-is while we figure out the name-mangled lookup on our end.

**Comment by joscha** [comment]
Thank you @sandersn ❤️ !

File: types/storybook__addon-knobs/index.d.ts (hunk 1 above)
Comment (andy-ms) [inline types/storybook__addon-knobs/index.d.ts]: See #17434.

File: types/storybook__addon-knobs/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    function boolean(name: string, value: boolean): boolean;
+
+    function number(name: string, value: number, options?: NumberOptions): number;
+
+    function color(name: string, value: string): string;
+
+    function object<T>(name: string, value: T): T;
+
+    function select<T>(name: string, options: { [s: string]: T }, value: string): T;
```
Comment (FDiskas) [inline types/storybook__addon-knobs/index.d.ts]: There should be one more argument "groupName"

Comment (andy-ms) [reply types/storybook__addon-knobs/index.d.ts]: PRs accepted.

Comment (joscha) [reply types/storybook__addon-knobs/index.d.ts]: Opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24797

**Pull request merged**